### PR TITLE
USHIFT-247: Generate release-info

### PR DESCRIPTION
**Pull request by fzdarsky**
Adds a new RPM package "microshift-release-info" that contains JSON files containing the container image digests of the respective MicroShift release, so users can use these digests when creating osbuild-composer blueprints.

As side-effect, moves MicroShift's own release info from the pkg/release/release_*.go into JSON files in the assets/release directory.

Signed-off-by: Frank A. Zdarsky <fzdarsky@redhat.com>

Closes [USHIFT-247](https://issues.redhat.com//browse/USHIFT-247)

### CODE REVIEW

**Comment by ggiguash** [comment]
@fzdarsky, can you add a short doc for devs explaing how this information can be used out of the new RPM?

File: packaging/rpm/microshift.spec
Comment (ggiguash) [inline packaging/rpm/microshift.spec:258]: Is it not 4.13 change?

File: docs/rhel4edge_iso.md
Comment (ggiguash) [inline docs/rhel4edge_iso.md:169]:
```suggestion
VERSION=$(grep "^var Base =" ~/microshift/pkg/release/release.go | cut -d\" -f 2)
```

File: pkg/release/release.go
Comment (ggiguash) [inline pkg/release/release.go:53]: Is it a good time to remove generic specs? Are they any good?

File: pkg/release/release.go
Comment (ggiguash) [inline pkg/release/release.go:68]: Can we add some comments in this section on what we're trying to do and why?

File: scripts/image-builder/build.sh
Comment (ggiguash) [inline scripts/image-builder/build.sh:7]:
```suggestion
IMAGE_VERSION=$(grep "^var Base =" "${ROOTDIR}/pkg/release/release.go" | cut -d\" -f 2)
```

**Review by ggiguash — CHANGES_REQUESTED** [review]

**Comment by ggiguash** [comment]
@fzdarsky, please update [this doc section](https://github.com/openshift/microshift/blob/409d4e6e42f3c69e8a45b68c099f2b7966aa864a/docs/devenv_rhel8.md?plain=1#L93) with the new RPM

File: packaging/rpm/microshift.spec
Comment (ggiguash) [inline packaging/rpm/microshift.spec:64]: Should we add a dependency on the new RPM so that it's automatically deployed when "dnf install -y microshift" command is run?

**Review by ggiguash — CHANGES_REQUESTED** [review]

File: packaging/rpm/microshift.spec
Comment (fzdarsky) [reply packaging/rpm/microshift.spec:64]: Well, the problem is this information isn't needed on the machine where MicroShift is run, it is needed on the machine where osbuild-composer blueprints are updated (which may not even be an RPM-based machine, so this solution is probably just temporary). So I don't think adding this a required dependency (even if small) is relevant.

File: pkg/release/release.go
Comment (fzdarsky) [reply pkg/release/release.go:53]: In the upstream MicroShift project, this has been our fallback for cases when no equivalent images from OCP exist that we could reference. My assumption is we'll needed when we revive the upstream community.

File: packaging/rpm/microshift.spec
Comment (ggiguash) [reply packaging/rpm/microshift.spec:64]: I would like to see a document on the usage scenarios before I make up my mind about this.

File: pkg/release/release.go
Comment (fzdarsky) [reply pkg/release/release.go:70]: Done.

**Comment by fzdarsky** [comment]
> @fzdarsky, please update [this doc section](https://github.com/openshift/microshift/blob/409d4e6e42f3c69e8a45b68c099f2b7966aa864a/docs/devenv_rhel8.md?plain=1#L93) with the new RPM

Done.

File: pkg/release/release.go
Comment (oglok) [inline pkg/release/release.go:47]: I wonder why arch names have been replaced: `amd64` to `x86_64` and `arm64` to `aarch64`

**Comment by oglok** [comment]
I think it's a very smart approach to solve this problem.

File: packaging/rpm/microshift.spec
Comment (dhellmann) [reply packaging/rpm/microshift.spec:64]: The contents of the files going into this new RPM are also compiled into the microshift binary like the other assets, so microshift does not need the files to be installed. It would not be wrong to install them on the host running microshift, but nothing would necessarily consume them.

**Review by dhellmann — COMMENTED** [review]
This is pretty much exactly what I had in mind. I had one or two nits, but nothing to hold it up. I know there were some other commenters, so I'll leave a hold along with the approval and you can remove that when everyone else is satisfied.

/lgtm
/hold

File: pkg/release/release.go
Comment (fzdarsky) [reply pkg/release/release.go:47]: My motivation for this was user experience. The x86_64/aarch64 architecture names are more idiomatic on Linux systems in general and RPM-based distros in particular.

For example, this way users can use `$(uname -i)` in scripts and avoid conditional logic and substitutions as with the Golang names. It’s also consistent with the RPM package naming scheme.

File: packaging/rpm/microshift.spec
Comment (ggiguash) [reply packaging/rpm/microshift.spec:64]: Still, a document is missing in this PR. We need to document in GitHub what a dev can do with this RPM. A mostly copy-paste from the proposal doc?

**Comment by ggiguash** [comment]
@fzdarsky, can we please implement 2 changes before we merge this PR?
- Add a document in GitHub (referenced from the README.md) explaining the RPM usage scenarios
- Restore the get.sh script and fix it to support the new parsing.

File: scripts/image-builder/build.sh
Comment (ggiguash) [reply scripts/image-builder/build.sh:7]: I agree. I think removing get.sh script forced us to duplicate and spread parsing of this information in other scripts. Let's return the get.sh file

File: pkg/release/release.go
Comment (mangelajo) [inline pkg/release/release.go:50]: It would be great if we could also produce a toml snippet so customers just can append that to their blueprints without additional processing if they wanted.

I suggest that as follow up, not now.

File: packaging/rpm/microshift.spec
Comment (mangelajo) [reply packaging/rpm/microshift.spec:64]: Yes, we should not add a dependency, but documentation is always good.

File: packaging/rpm/microshift.spec
Comment (mangelajo) [inline packaging/rpm/microshift.spec:72]: May be we should explain the purpose of this; i.e.

This package can be used to find the images consumed by MicroShift in runtime
and embed those into blueprints/ostree images.

**Comment by fzdarsky** [comment]
> @fzdarsky, can you add a short doc for devs explaing how this information can be used out of the new RPM?

Added new `docs/howto_offline_containers.md` and linked it.

File: scripts/image-builder/build.sh
Comment (fzdarsky) [reply scripts/image-builder/build.sh]: Moved the base into the JSON, too, so now parsing is uniform and simple.

File: packaging/rpm/microshift.spec
Comment (fzdarsky) [reply packaging/rpm/microshift.spec]: Added.

**Comment by mangelajo** [comment]
/lgtm

Still get.sh is not re-added as talked in slack. But we can add that back if necessary.

**Pull request merged**
